jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java

```diff
@@ -1229,36 +1237,52 @@ public UnavailableServlet(UnavailableException unavailableException, Servlet ser
             }
             else
             {
-                _servlet = servlet;
-                _available = System.nanoTime() + TimeUnit.SECONDS.toNanos(unavailableException.getUnavailableSeconds());
+                _unavailableServlet = servlet;
+                long start = System.nanoTime();
+                while (start == 0)
+                    start = System.nanoTime();
+                _unavailableStart = new AtomicLong(start);
             }
         }
 
         @Override
         public void service(ServletRequest req, ServletResponse res) throws ServletException, IOException
         {
-            if (_available == -1)
-                ((HttpServletResponse)res).sendError(HttpServletResponse.SC_NOT_FOUND);
-            else if (System.nanoTime() < _available)
-                ((HttpServletResponse)res).sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE);
-            else
+            while (true)
```

Comment:
Can we come up with a solution that removes the while(true) - I think these are best avoided if possible.

Reply:
Will do, because I can in this situation.  However in general, the while(true) pattern is the best and simplest approach for atomic state machines.   More often than not, we've been wrong when we do that.